roomserver/storage/shared/storage.go

```diff
@@ -823,13 +823,38 @@ func (d *Database) handleRedactions(
 		return nil, "", nil
 	}
 
+	// Get the power level from the database, so we can verify the user is allowed to redact the event
+	powerLevels, err := d.GetStateEvent(ctx, event.RoomID(), gomatrixserverlib.MRoomPowerLevels, "")
+	if err != nil {
+		return nil, "", fmt.Errorf("d.GetStateEvent: %w", err)
+	}
+	pl, err := powerLevels.PowerLevels()
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to get powerlevels for room: %w", err)
+	}
+	redactPL := pl.Redact
+	redactUser := pl.Users[redactionEvent.Sender()]
+	// The power level of the redaction event’s sender is greater than or equal to the redact level.
+	userAllowed := redactUser >= redactPL
+	// The domain of the redaction event’s sender matches that of the original event’s sender.
+	originAllowed := redactedEvent.Origin() == redactionEvent.Origin()
```

Comment:
Oh, yes, that wasn't first-glance-friendly. May be easier to parse as:
```
select {
case redactUser >= redactPL:
  // The power level of the redaction event's sender is greater than or equal to the redact level.
case redactedEvent.Origin() == redactionEvent.Origin():
  // The domain of the redaction event’s sender matches that of the original event’s sender.
default:
  return nil, "", nil
}
```
... which also makes me wonder whether we should be checking both `Sender` and `Origin` (like `redactedEvent.Origin() == redactionEvent.Origin() && redactedEvent.Sender() == redactionEvent.Sender()`), otherwise it seems like it would allow another user on the same homeserver to send redactions.

Reply:
Right, but that (at least in theory) should already be denied on the other homeserver, given the powerlevels.
Looking at Sytest, it only tests with sending a bogus redaction event from a different user/homeserver, and a legitimate redaction of the sender of the redacted event.
Checking only `redactedEvent.Sender() == redactionEvent.Sender()` (which is allowed anyway) also passes the test.

Reply:
> Right, but that (at least in theory) should already be denied on the other homeserver, given the powerlevels

True, but we want to make sure we aren't accepting things when malicious actors are lying to us, so better to be safe and thorough.